File: Knou/2023년 2학기/simulation/java_src(file)/EX3_2.java
/*--------------------------------------------------------------
 * 봉사시간 변화 시뮬레이션
 * TLIMIT : 총 시뮬레이션 수행 시간
 * TIME   : 시뮬레이션의 현재시간(단위 : 분)
 * TSTEP  : 시뮬레이션 진행 단위 시간
 * TPUMP  : 고객에 대한 봉사시간(단위 : 분)
 * PRARR  : 고객이 1분내에 도착할 확률
 * MEAN   : 평균으로 실수형 변수
 * STDEV  : 표준편차를 의미하는 실수형 변수
 * AVEQUE : 대기행렬 queue 의 평균길이
 * TOTQUE : 대기행렬 queue 의 전체 합 (＝전체 고객의 총 대기시간)
 * TOTARR : 도착한 총 고객수
 *-------------------------------------------------------------*/
import java.lang.*;
import java.io.*;

class Queueing {
      final static int SEED = 35213;
      static float up, liters;
      public String StrTime = new String();

      public int random(int np, double u){
              np = np * 843314861 + 453816693;
              if(np < 0){
                      np = np + 2147483647;
                      np = np + 1;
              }
              up = (float)(np * 0.4656612e-9);
              return np;
      }

      public int normal(int np, float mean, float std, float lp){
              int i = 0;
              float sum = 0, u;

              while(i < 12) {
                      np = random(np, up);
                      sum = sum + up;
                      ++i;
              }
              liters = mean + (sum - 6.0f) * std;
              return np;
     }

     public void ConvDataToString(int time){
            int i, len;

            StrTime = "";
            StrTime = StrTime.valueOf(time);
            len = StrTime.length();
            if(len < 3)
                 for(i = 0; i < (3-len); i++)
                      StrTime = "0" + StrTime;
     }

      public void ComputeQueueing(){
              int seed, nseed, queue = 0, totque = 0, totarr=0, arrive, tstep = 1;
              float nmean = 20, stdev = 3;
              double prarr=1.0/4.0, tpump=0.0, tlimit=100, time=0.0, aveque=0;

              seed = SEED;
              nseed = SEED;

              PrintWriter out=null;

              try {
                  File f=new File("C:\\WORK\\EX3_2.OUT");
                  FileWriter fw=new FileWriter(f) ;
                  BufferedWriter bw=new BufferedWriter(fw);
                  out=new PrintWriter(bw);

                  out.println("   SIMULATION FOR A QUEUEING SYSTEM ");
                  out.println("======================================");
                  out.println(" THE TIME STEP                 = " + tstep);
                  out.println(" THE TIME LIMIT                = " + (int)tlimit);
                  out.println(" THE ARRIVAL PROBABILITY       = " + (double)((int)(prarr*100))/100);
                  out.println(" THE NORMAL MEAN               = " + (int)nmean);
                  out.println(" THE NORMAL STANDARD DEVIATION = " + (int)stdev);
                  out.println(" THE SEED                      = " + seed);
                  out.println("--------------------------------------");
                  out.println(" TIME   ARRIVAL   QUEUE    TPUMP");
                  out.println("--------------------------------------");

                  while(time < tlimit) {
                          time = time + tstep;
                          arrive = 0;
                          seed = random(seed, up);
                          if(up < (prarr*tstep)) {
                                  arrive = 1;
                                  queue = queue + arrive;
                                  totarr = totarr + 1;
                          }
                          if(tpump > 0.0) {
                                  tpump = tpump - tstep;
                                  if(tpump < 0.0)
                                        tpump = 0.0;
                          }
                          if(tpump==0 && queue!=0) {
                                  queue = queue - 1;
                                  nseed = normal(nseed, nmean, stdev, liters);
                                  tpump = 1.0 + 1.0 / 8.0 * liters;
                          }
                          totque = totque + queue;
                          ConvDataToString((int)time);
                          out.println(" " + StrTime + "       " + arrive + "        " + queue + "      " + (double)(((int)(tpump*10000000))/10000000.0));
                  }
                  aveque = totque / (tlimit*tstep);
                  out.println("------------------------------------");
                  out.println(" MEAN QUEUE LENGTH = " + aveque);
                  out.println(" THE TOTAL ARRIVAL = " + totarr);
                } catch(IOException ioe) { }
                  finally {
                       if (out != null) out.close();
                  }
      }
}

public class EX3_2 {
        public static void main(String[] args){
                Queueing g = new Queueing();
                g.ComputeQueueing();
        }
}

File: Knou/2023년 2학기/시뮬레이션/VC++2019/ex32/ex3-2.c
// SIMULATION FOR A QUEUEING SYSTEM                               //
// TO EVALUATE MEAN QUEUE LENGTH                                  //
// -- EACH CAR HAS NORMAL DISTRIBUTION TANK --                    //

// TLIMIT   : the time limit expressed in minutes                 //
// TIME     : the time expressed in minutes                       //
// TSTEP    : the time increment in minutes                       //
// TPUMP    : the service time that a custom has left             //
// PRARR    : the probability of a custom arriving in any minute  //
// MEAN     : the mean of normal distribution                     //
// STDEV    : the standard deviation of normal distribution       //
// AVEQUE   : the average queue length                            //
// TOTQUE   : the totals of queue length                          //
// TOTARR   : the totals of arrivals                              //

// PROGRAM  EX3-2                                                 //

#include<stdio.h>
#include<math.h>
#define SEED 35213;


//  RANDOM NUMBER GENERATOR  //
void random(long *np, float *up)
{
       *np = *np * 843314861 + 453816693;
       if(*np < 0)
	      { *np = *np +2147483647;
		*np = *np + 1;
	      }
       *up = *np * 0.4656612e-9;
}


//  NORMAL RANDOM VARIABLE GENERATOR  //
void normal(long *np, float mean, float std, float *lp)
{
       int i=0;
       float sum=0,u;

       while ( i < 12 )
	    {
	      random(np,&u);
	      sum = sum + u;
	      ++i;
	    }
       *lp = mean + (sum - 6.0) * std;
}


//  MAIN  PROGRAM  //
void main()
{
       long seed,nseed;
       int   queue=0, totque=0, totarr=0, arrive, tstep=1;
       float nmean=20,stdev=3;
       float liters, prarr=1.0/4.0, tpump=0.0, time=0.0, tlimit=100, u=0, aveque=0;
       FILE *output;

	   seed=SEED;
	   nseed=SEED;
	   output=fopen("c:\\work\\ex3-2.out","wt");

       fprintf(output,"        SIMULATION FOR A QUEUEING SYSTEM \n\n");
       fprintf(output,"THE TIME STEP                  =  %d \n",tstep);
       fprintf(output,"THE TIME LIMIT                 =  %3.0f \n",tlimit);
       fprintf(output,"THE ARRIVAL PROBABILITY        =  %4.2f \n",prarr);
       fprintf(output,"THE NORMAL MEAN                =  %2.0f \n",nmean);
       fprintf(output,"THE NORMAL STANDARD DEVIATION  =  %1.0f \n",stdev);
       fprintf(output,"THE SEED                       =  %ld \n",seed);
       fprintf(output,"  TIME       ARRIVAL      QUEUE            TPUMP\n");

       while (time < tlimit)
	    {
	      time = time + tstep;
	      arrive = 0;
	      random(&seed, &u);
	      if (u < (prarr*tstep))
		    { arrive = 1;
		      queue = queue + arrive;
		      totarr = totarr + 1;
		    }
	      if (tpump > 0.0)
		    { tpump = tpump - tstep;
			  if (tpump < 0.0) tpump = 0.0;
		    }
	      if ((tpump == 0) && (queue != 0))
		    { queue = queue - 1;
		      normal(&nseed,nmean,stdev,&liters);
		      tpump = 1.0 + 1.0 / 8.0 * liters;
		    }
	      totque = totque + queue;
	      fprintf(output,"  %3.0f           %d          %2d           %10.7f \n",time,arrive,queue,tpump);
	    }
       aveque = totque / (tlimit/tstep);
       fprintf(output,"  MEAN QUEUE LENGTH  =  %4.2f \n",aveque);
       fprintf(output,"  THE TOTAL ARRIVALS =  %d \n",totarr);
       fprintf(output,"  THE TOTAL QUEUE =  %d \n",totque);
       fclose(output);
}
